### PR TITLE
WP/JPAndroid: Update Fastfile with version model

### PR DESCRIPTION
**Pull request by spencertransier**
## Summary
- This PR integrates the new version and build code models from https://github.com/wordpress-mobile/release-toolkit/pull/512
- Nearly all of the `precheck` actions from the Release Toolkit were tightly coupled with the existing version methods. To simplify things with the new version model, I took the relevant snippets from the precheck actions and added them directly here. These changes are somewhat opposite of DRY, but I think it's worth it (for now) for transparency and readability. We may also decide in the future to create smaller actions that do more specific things versus the existing precheck and version bump actions that combine many disparate methods.

### CODE REVIEW

File: fastlane/lanes/release.rb
Comment (jkmassel) [inline fastlane/lanes/release.rb]: Should this be the same as what we have on line 41?

**Review by jkmassel — CHANGES_REQUESTED** [review]
I smoke tested this and it appears to work, but let's coordinate with @oguzkocer on landing it?

File: fastlane/lanes/release.rb
Comment (spencertransier) [reply fastlane/lanes/release.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/pull/19288/commits/0283d84ea2d2002df1bb8e56c6bebe3b3107b211

**Comment by oguzkocer** [comment]
@spencertransier I haven't been following along with the version model changes, but if Jeremy approves the PR, feel free to merge it any time. I am guessing a second review is not necessary, but if it is, please let me know and I'll review it. I don't think I can do it today though.

**Comment by spencertransier** [comment]
@oguzkocer I think it would be helpful for you to take a look at it to make sure that none of the changes interfere with your work on CI release management. These changes are essentially just adding the Version model functionality by pulling out the functionality from the precheck and version bumper classes and bringing them directly into the Fastfile (they were so tightly coupled with the previous versioning logic that it was necessary to do it like that). 

That's all to say that it'd be great if you could give it a glance to see any potential issues for CI releases, when you have a chance 😃

**Comment by oguzkocer** [comment]
Sure thing - I'll try to take a look tomorrow.

File: fastlane/Fastfile
Comment (oguzkocer) [inline fastlane/Fastfile]: Should we use different variable names to avoid confusion? Maybe we can name the first one as `next_minor_version`?

File: fastlane/Fastfile
Comment (oguzkocer) [inline fastlane/Fastfile:121]: What happens if this is called for a production version - meaning the next beta version is the first one?

File: fastlane/Fastfile
Comment (oguzkocer) [inline fastlane/Fastfile]: Why are we moving this from `before_all` lanes to `code_freeze`? More importantly, why are we doing it in this PR?

Furthermore, unless I am missing something, `code_freeze` is the lane we **wouldn't** want this check, because we can't actually do the update without creating a PR.

File: fastlane/lanes/build.rb
Comment (oguzkocer) [inline fastlane/lanes/build.rb]: Since this message is printed regardless of the value of `skip_confirm`, should we use `UI.message` to print this message out and then only do the confirmation if the `skip_confirm` is `false`?

File: fastlane/lanes/build.rb
Comment (oguzkocer) [inline fastlane/lanes/build.rb]: Should we move this to the previous `unless options[:skip_prechecks]` block?

File: fastlane/lanes/build.rb
Comment (oguzkocer) [inline fastlane/lanes/build.rb]: This is still using the `version` variable which no longer exists.

File: fastlane/lanes/build.rb
Comment (oguzkocer) [inline fastlane/lanes/build.rb]: Should we rename `version` to `version_name` to make it more clear what this needs? `version` is a very loaded name and hard to know what it's referring to - especially in a language that doesn't have a type system.

This comment might be applicable in other places as well.

File: fastlane/lanes/release.rb
Comment (oguzkocer) [inline fastlane/lanes/release.rb]: Similar to a previous comment, it looks like we can always print out the message and then conditionally do the confirmation.

File: fastlane/lanes/release.rb
Comment (oguzkocer) [inline fastlane/lanes/release.rb]: I can't find this method in WPAndroid or `release-tooltkit` :( could you help me with that?

File: fastlane/lanes/release.rb
Comment (oguzkocer) [inline fastlane/lanes/release.rb]: I don't think we are using `base_version` option anywhere to trigger this lane. Should we use this opportunity to drop it?

File: fastlane/lanes/release.rb
Comment (oguzkocer) [inline fastlane/lanes/release.rb]: Similar to a previous comment, it looks like we can always print out the message and then conditionally do the confirmation.

File: fastlane/lanes/release.rb
Comment (oguzkocer) [inline fastlane/lanes/release.rb]: Should we do this after `Fastlane::Helper::GitHelper.checkout_and_pull(release: current_release_version)`? If we are going to do this verification, we might as well do it before updating frozen strings and downloading translations.

File: fastlane/lanes/release.rb
Comment (oguzkocer) [inline fastlane/lanes/release.rb]: Similar to a previous comment, it looks like we can always print out the message and then conditionally do the confirmation.

File: fastlane/lanes/release.rb
Comment (oguzkocer) [inline fastlane/lanes/release.rb]: Similar to a previous comment, should we rename `version` to `version_name`? I think we want to do this everywhere we were using `version['name']` because otherwise it's going to be very confusing what `version` refers to.

File: fastlane/lanes/release.rb
Comment (oguzkocer) [inline fastlane/lanes/release.rb]: Similar to a previous comment, it looks like we can always print out the message and then conditionally do the confirmation.

File: fastlane/lanes/release.rb
Comment (oguzkocer) [inline fastlane/lanes/release.rb]: Should we rename this to `version_name` to differentiate it from the previous `version` model?

**Review by oguzkocer — CHANGES_REQUESTED** [review]
@spencertransier I've completed my first pass, I left some feedback/questions. I hope you find them useful.

---

I find these reviews to be incredibly difficult. There is no type system in Ruby and I can't test the changes and that makes it very hard to verify the changes. I ran one of the lanes and that one failed and I imagine this is likely to happen in other lanes even after we merge the changes. So, I focused on trying to understand the code with the goal of being able to quickly address any issues that might arise during the next release cycle.

I am still not super clear on how the new versioning scheme works as I didn't review that PR, but I see that it's being merged into several projects, so I guess it's already verified? If we are not fully confident in it, we might want to stagger the merges so we can address issues we come across in other projects.

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: My thinking was that it's better to update the Release Toolkit earlier in the code freeze than later so that any changes can be tested. So having it check before every lane is run doesn't seem efficient. But I see your point and will add it back to `before_all`.

Comment (oguzkocer) [reply fastlane/Fastfile]: If we want to go with that route, I think `complete_code_freeze` would make the most sense. While running that lane, we are already in the `release` branch and it's a decent time to do the update. However, if we want to do that, I think it should be done for all apps after consensus amongst the team. We should also separate it into its own PR since it is not related to this one.

File: fastlane/lanes/release.rb
Comment (spencertransier) [reply fastlane/lanes/release.rb]: @oguzkocer Do you ever use it for hotfixes? I seem to recall needing to use it on other apps once in a while. Maybe I'm misremembering. But yep, I am definitely open to removing it if you don't think we need it anymore.

Comment (oguzkocer) [reply fastlane/lanes/release.rb]: Yeap, we use a base version for the hotfix, but this is the `new_beta_release` lane, so I don't think it'll matter?

File: fastlane/lanes/release.rb
Comment (spencertransier) [reply fastlane/lanes/release.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6

File: fastlane/lanes/release.rb
Comment (spencertransier) [reply fastlane/lanes/release.rb]: This is now added at the end of `release.rb`. 

Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6

File: fastlane/lanes/release.rb
Comment (spencertransier) [reply fastlane/lanes/release.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6

File: fastlane/lanes/release.rb
Comment (spencertransier) [reply fastlane/lanes/release.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6

File: fastlane/lanes/release.rb
Comment (spencertransier) [reply fastlane/lanes/release.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6

File: fastlane/lanes/release.rb
Comment (spencertransier) [reply fastlane/lanes/release.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6

File: fastlane/lanes/release.rb
Comment (spencertransier) [reply fastlane/lanes/release.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6

File: fastlane/lanes/build.rb
Comment (spencertransier) [reply fastlane/lanes/build.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6

File: fastlane/lanes/build.rb
Comment (spencertransier) [reply fastlane/lanes/build.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6

File: fastlane/lanes/build.rb
Comment (spencertransier) [reply fastlane/lanes/build.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6

File: fastlane/lanes/build.rb
Comment (spencertransier) [reply fastlane/lanes/build.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: I moved it back to `before_all` in https://github.com/wordpress-mobile/WordPress-Android/commit/6ab18efe9261a63b77766614c06c3bc661a697f6 so we can discuss it with the team before making any changes here.

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile:121]: @oguzkocer Sorry, I'm a bit confused about this one. Could you give me an example?

File: fastlane/lanes/release.rb
Comment (spencertransier) [reply fastlane/lanes/release.rb]: @oguzkocer I was thinking that there could be a circumstance where a new hotfix build is needed, but you're right--because the version is bumped before hotfix PRs are merged into the branch and a build isn't kicked off until after that, this isn't needed.

File: fastlane/lanes/release.rb
Comment (spencertransier) [reply fastlane/lanes/release.rb]: I decided to remove this check because we're already checking that the current version's release branch is being checked out here: 

```
unless Fastlane::Helper::GitHelper.checkout_and_pull(release: current_release_version)
      UI.user_error!("Release branch for version #{current_release_version} doesn't exist.")
    end
```

Comment (spencertransier) [reply fastlane/lanes/release.rb]: Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/08c57ebfd46cef2dc7a0f4c8943fc58abc5f0d86

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Addressed in Addressed in https://github.com/wordpress-mobile/WordPress-Android/commit/08c57ebfd46cef2dc7a0f4c8943fc58abc5f0d86

**Comment by spencertransier** [comment]
@oguzkocer Ready for re-review 🙂

File: fastlane/Fastfile
Comment (oguzkocer) [inline fastlane/Fastfile]: After the change in 08c57ebfd46cef2dc7a0f4c8943fc58abc5f0d86, I _think_ this needs to change like so:

```suggestion
  first_beta_version = VERSION_CALCULATOR.next_build_number(version: next_minor_version)
```

File: fastlane/lanes/build.rb
Comment (oguzkocer) [inline fastlane/lanes/build.rb:19]: This is not related to this PR since this is just moved from elsewhere, but I do wonder if we should really hardcode `is_ci` checks like this - especially since we already have a way to skip these checks in CI by passing `skip_prechecks`. My worry is that hardcoding checks like this makes it harder to reason about the callers without looking at the source code.

@AliSoftware we recently had a discussion about the "is CI" checks in Tumblr and I'd love to hear what you think about this in general.

P.S: I think this is a good place to bring up this discussion, but let's not make any changes to this PR related to this. If we decide removing the checks, we can do that in its own PR.

Comment (AliSoftware) [reply fastlane/lanes/build.rb:19]: Actually a wider question came up as part of @spencertransier 's work on the versioning model.

In particular, the various `*_prechecks` actions that we have in the `release-toolkit` are (1) kinda obscure in what they're doing/checking (2) too tied with assumptions about the version model and format used by the projects (3) should probably be reworked or broken into more atomic and specific actions (similar to calling the more atomic `ensure_git_branch` here instead of calling some obscure `android_*_prechecks` action).

So we might want to have a broader discussions about those prechecks in general.

To answer your question directly though, I agree that it makes more sense to just have a `options[:skip_prechecks]` test wrapping precheck code, and not using `unless is_ci`. We can then always call the lane with `skip_prechecks:true` within the `.buildkite/commands/*.sh` invoking that lane from CI steps if we want, which would make things more explicit and avoid having redundant flags.

File: fastlane/Fastfile
Comment (oguzkocer) [reply fastlane/Fastfile:121]: I see that we have `first_beta_version` and `next_beta_version` helpers. I was asking what happens if we accidentally call `next_beta_version` when we should have been calling `first_beta_version`. Is that anything concerning that we should try?

I _think_ `first_beta_version` is only supposed to be called from `code_freeze` and we are already doing that, so this is unlikely to be an issue. However, I thought it was still worth bringing up as a question.

**Review by oguzkocer — CHANGES_REQUESTED** [review]
@spencertransier I left [one comment that needs attention](https://github.com/wordpress-mobile/WordPress-Android/pull/19288/files#r1357034027) and replied to your question [here](https://github.com/wordpress-mobile/WordPress-Android/pull/19288/files#r1357110050). Other than that I think this one is good to merge in time for the next code freeze.

It's likely that we missed some minor things, but I think I'll be able to address them as part of the next release cycle.

File: fastlane/lanes/build.rb
Comment (oguzkocer) [reply fastlane/lanes/build.rb:19]: > (3) should probably be reworked or broken into more atomic and specific actions

That sounds like a great plan to me!

Comment (spencertransier) [reply fastlane/lanes/build.rb:19]: >  do wonder if we should really hardcode `is_ci` checks like this - especially since we already have a way to skip these checks in CI by passing `skip_prechecks`

I had the same thought while breaking out these `*_prechecks` actions into the app Fastfiles. These hardcoded `is_ci` checks existed in the `*_prechecks` actions, but I didn't have any context for why they were there in the first place or why they are added to the same check in some of the precheck actions but not others. So overall, I agree that removing the hardcoded `is_ci`'s and moving them them to the `skip_prechecks` section is a great idea.

>  (3) should probably be reworked or broken into more atomic and specific actions

This is something I'm planning to work on after the env var project is finished 👍

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Good catch, thank you!

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile:121]: @oguzkocer Ah, I see what you mean. Yes, the `first_beta_version` should only be used during the code freeze. I updated the name from `first_beta_version` to `code_freeze_beta_version` in https://github.com/wordpress-mobile/WordPress-Android/pull/19288/commits/65dbfaad62d3c765103a41ea480400216f075884 to clarify that.

**Comment by spencertransier** [comment]
@oguzkocer I applied the commit you suggested and addressed your `first_beta_version` comment here: https://github.com/wordpress-mobile/WordPress-Android/pull/19288#discussion_r1357197302

**Review by oguzkocer — APPROVED** [review]
Thanks for the PR and the follow up updates @spencertransier 🙇

**Review by jkmassel — APPROVED** [review]
My suggestions have been addressed

**Pull request merged**
